### DISCUSSION TITLE
Recursive graph

### DISCUSSION THREAD

**Discussion by chrishiste** [dagster-io/dagster#22472] [Q&A]
I have a graph that extracts a dataset in batches. The graph consists of a few distinct operations to create all the batches and extract them using a Dynamic Output.

Sometimes, the extraction of a single batch fails because its content is too large to extract in one attempt. I would like to implement a mechanism to recursively break down that batch into smaller batches using the same logic as the parent graph. This process should continue recursively, reducing the batch size each time, until the entire graph completes successfully.

**Problem:**

- The current graph occasionally fails to extract a batch due to its large size.
- There is no way to precompute the size of each batch in advance to know if it will fail. 
- I need a mechanism to retry the extraction on failure and reduce the batch size dynamically.
- The logic used by the parent graph to extract batches should be applied to the smaller batches.

**Example:**

1. Initial Extraction Attempt:

- Graph attempts to extract Batch A.
- Extraction fails due to large size.

2. First Recursive Reduction:

- Batch A is split into smaller Batches A1, A2, A3, etc.
- Graph attempts to extract each smaller batch.

3. Subsequent Reductions (if necessary):

- If Batch A1 fails, it is further split into A1a, A1b, A1c, etc.
- This process continues until all smaller batches are successfully extracted.

Is there currently a way to do such recursive operations? I couldn't find a way to configure a retry policy on subgraphs. I could hack my way around this problem using run retries but unfortunately they are not compatible with asset_graph at the moment.

**Comment by chrishiste** [comment]
@garethbrickman any chance we could get someone to look into this?

**Comment by garethbrickman** [comment]
I don't think there's a way to run an exception recursively in an op the way you're envisioning. As you point out [in your other discussion](https://github.com/dagster-io/dagster/discussions/23544), this would require try-catch type functionality in graphs for fallback handling. 

Outside of using retry policies, the other idea that came to mind would be to perform the batch extraction externally either as an [External Asset](https://github.com/dagster-io/dagster/discussions/23544) or using [Dagster Pipes](https://docs.dagster.io/concepts/dagster-pipes) to separate the orchestration from the business logic.

**Reply by chrishiste** [reply]
Thanks, External Asset and Dagster Pipes don't seem to fit my use case. I think what I'm lacking in Dagster are advanced state machine capabilities. Dynamic graphs are amazing, but I would really benefit from a way to incorporate more robust state transitions and error handling directly within Dagster graphs.

**Reply by chrishiste** [reply]
Allowing us to incorporate Python features like conditional logic, loops, and try-except blocks directly within Dagster graphs would make this possible.

**Reply by chrishiste** [reply]
Prefect seems to allow that
![ScreenShot 2024-08-14 at 21 32 01@2x](https://github.com/user-attachments/assets/e5fcd1f2-6fa5-47c9-b7fa-a3b2ed8877d3)
